src/messaging/tests/TestReliableMessageProtocol.cpp

```diff
@@ -0,0 +1,271 @@
+{
+    TestContext & ctx = *reinterpret_cast<TestContext *>(inContext);
+    auto & m          = ReliableMessageManager::GetManager();
+    m.TestSetIntervalShift(4); // 16ms per tick
+    m.Init(ctx.GetSystemLayer());
+    ReliableMessageContext rc;
+    ReliableMessageDelegateObject delegate;
+    rc.SetDelegate(&delegate);
+    rc.SetConfig({
+        1, // CHIP_CONFIG_RMP_DEFAULT_INITIAL_RETRANS_TIMEOUT_TICK
+        1, // CHIP_CONFIG_RMP_DEFAULT_ACTIVE_RETRANS_TIMEOUT_TICK
+        1, // CHIP_CONFIG_RMP_DEFAULT_ACK_TIMEOUT_TICK
+        1, // CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS
+    });
+    ReliableMessageManager::RetransTableEntry * entry;
+    auto buf = System::PacketBuffer::New();
+    m.AddToRetransTable(&rc, buf, 1, 0, &entry);
+    m.ProcessDelayedDeliveryMessage(&rc, 64);
+    NL_TEST_ASSERT(inSuite, m.TestGetCountRetransTable() == 1);
+
+    test_os_sleep_ms(50);
+    ReliableMessageManager::Timeout(&ctx.GetSystemLayer(), &m, CHIP_SYSTEM_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, m.TestGetCountRetransTable() == 1);
+    // not send, delayed
+
+    test_os_sleep_ms(50);
+    ReliableMessageManager::Timeout(&ctx.GetSystemLayer(), &m, CHIP_SYSTEM_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, m.TestGetCountRetransTable() == 1);
+    NL_TEST_ASSERT(inSuite, !delegate.SendErrorCalled);
+    // 1st retrans
+
+    test_os_sleep_ms(20);
+    ReliableMessageManager::Timeout(&ctx.GetSystemLayer(), &m, CHIP_SYSTEM_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, m.TestGetCountRetransTable() == 0);
+    NL_TEST_ASSERT(inSuite, delegate.SendErrorCalled);
+    // send error
+
+    m.Shutdown();
+}
+
+// Test Suite
+
+/**
+ *  Test Suite that lists all the test functions.
+ */
+// clang-format off
+const nlTest sTests[] =
+{
+    NL_TEST_DEF("Test ReliableMessageManager::CheckAddClearRetrans", CheckAddClearRetrans),
+    NL_TEST_DEF("Test ReliableMessageManager::CheckFailRetrans", CheckFailRetrans),
+    NL_TEST_DEF("Test ReliableMessageManager::CheckRetransExpire", CheckRetransExpire),
+    NL_TEST_DEF("Test ReliableMessageManager::CheckDelayDelivery", CheckDelayDelivery),
+
+    NL_TEST_SENTINEL()
+};
+// clang-format on
+
+int Initialize(void * aContext);
+int Finalize(void * aContext);
+
+// clang-format off
+nlTestSuite sSuite =
+{
+    "Test-CHIP-ReliableMessageProtocol",
+    &sTests[0],
+    Initialize,
+    Finalize
+};
+// clang-format on
+
+/**
+ *  Initialize the test suite.
+ */
+int Initialize(void * aContext)
+{
+    CHIP_ERROR err = reinterpret_cast<TestContext *>(aContext)->Init(&sSuite);
+    return (err == CHIP_NO_ERROR) ? SUCCESS : FAILURE;
+}
+
+/**
+ *  Finalize the test suite.
+ */
+int Finalize(void * aContext)
+{
+    CHIP_ERROR err = reinterpret_cast<TestContext *>(aContext)->Shutdown();
+    return (err == CHIP_NO_ERROR) ? SUCCESS : FAILURE;
+}
+
+} // namespace
+
+namespace chip {
+namespace messaging {
+
+// Stub implementation
+ReliableMessageManager & ReliableMessageManager::GetManager()
```

Comment:
We should try to remove this kind of coupling.

Reply:
done